### PR TITLE
Init first draft version of rocketmq new apis

### PR DESCRIPTION
**Pull request by zhouxinyu**
We have drafted the new rocketmq protocol and APIs, let's discuss it here or on the dev mailing list.

### CODE REVIEW

**Comment by duhenglucky** [comment]
Why don't we put this in the [rocketmq](https://github.com/apache/rocketmq) repo? It seems like a better option

File: apache/rocketmq/v1/definition.proto [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+// dead lettered multiple times.
+//
+// If validation on any of the fields fails at subscription creation/update,
+// the create/update subscription request will fail.
+message DeadLetterPolicy {
+  // The maximum number of delivery attempts for any message.
+  //
+  // This field will be honored on a best effort basis.
+  //
+  // If this parameter is 0, a default value of 5 is used.
```
Comment (lizhanhui) [inline apache/rocketmq/v1/definition.proto:49]: The default value here is incorrect. Default max attempt is 16

**Comment by zhouxinyu** [comment]
> Why don't we put this in the [rocketmq](https://github.com/apache/rocketmq) repo? It seems like a better option

OK, we can consider move it to the main repo or add it as a git submodule.

File: apache/rocketmq/v1/definition.proto (hunk 1 above)
Comment (zhouxinyu) [reply apache/rocketmq/v1/definition.proto]: Good catch, resolved~

File: README.md [hunk 2]
```diff
@@ -1 +1,13 @@
...
+
+## Message Model
+
+## Producer Flow
+
+## Consumer Flow
+
+## Reserved Fields
+
+This API spec has reserved the numeric value of PB enum to 64 and the field numeric value of PB message to 64. These reserved values will be used by RocketMQ's official implementation. Beyond the reserved values are free to use by external organizations.
```
Comment (lizhanhui) [inline README.md:13]: Messages and enumerations of the API spec reserve the first 64 fields for the evolution of RocketMQ exclusively. Vendor-specific extensions to the protocol are supposed to use fields beyond 64.

File: apache/rocketmq/v1/service.proto [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+}
+
+message NotifyClientTerminationResponse {
+  ResponseCommon common = 1;
+
+  reserved 2 to 64;
+}
+
+service MessagingService {
+  rpc QueryRoute(QueryRouteRequest) returns (QueryRouteResponse) {}
```
Comment (lizhanhui) [inline apache/rocketmq/v1/service.proto:350]: Service items should be properly documented, describing their usage, context, and more importantly, error handling.

A typical example for reference: https://github.com/envoyproxy/data-plane-api/blob/main/envoy/api/v2/discovery.proto

Comment (zhouxinyu) [reply apache/rocketmq/v1/service.proto:364]: in progress~

File: apache/rocketmq/v1/service.proto [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+  reserved 3 to 64;
+}
+
+message HealthCheckResponse {
+  ResponseCommon common = 1;
+
+  reserved 2 to 64;
+}
+
+message EndTransactionRequest {
```
Comment (drpmma) [inline apache/rocketmq/v1/service.proto:190]: It seems that tranStateTableOffset and commitLogOffset are not included in this protocol.

Comment (zhouxinyu) [reply apache/rocketmq/v1/service.proto:190]: resolved

**Comment by lizhanhui** [comment]
I'd merge this pull request and make it a base for further discussion and refinement.

**Pull request merged**